x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/ApplicationPermission.java

```diff
@@ -83,6 +83,29 @@ public boolean grants(ApplicationPrivilege other, String resource) {
         return matched;
     }
 
+    /**
+     * Determines if this {@link ApplicationPermission} is a subset of other
+     * application permission. It does this by checking whether each of the
+     * permissions for this is granted by the other application permissions.
+     *
+     * @param other application permission
+     * @return {@code true} if this is subset of other else it is {@code false}
+     */
+    public boolean isSubsetOf(final ApplicationPermission other) {
```

Comment:
From recollection when I implement get-user-privileges, this method isn't going to find as many subsets as you might expect.
Because permission automata (both privilege names and resources) get `union`ed together quite early on, things that look like they should be detected as a subset, won't be because they were unioned with something else.

See the comment on `getResourcePatterns` for an example of something that feels like it should be detected as a duplicate (subset) but can't be.

Not to say this isn't worth having, just pointing out that it will sometimes acheive less that we might expect.